csrc/abstract_tensor.h

```diff
@@ -253,70 +580,262 @@ struct AbstractTensor {
     return domain.crend();
   }
 
-  AbstractTensor& pushBack(AbstractId id) {
+  AbstractTensorWithInfo<Info>& pushBack(AbstractId id) {
     domain.push_back(std::move(id));
+    info.resize(domain.size());
+    return *this;
+  }
+
+  AbstractTensorWithInfo<Info>& pushBack(AbstractId id, const Info& info) {
+    domain.push_back(std::move(id));
+    info.push_back(info);
     return *this;
   }
 
   template <typename... Args>
-  AbstractTensor& emplaceBack(Args&&... args) {
+  AbstractTensorWithInfo<Info>& emplaceBack(Args&&... args) {
     domain.emplace_back(std::forward<Args>(args)...);
     return *this;
   }
 
   template <typename T>
   bool operator==(T&& t) const {
-    if constexpr (std::is_same_v<AbstractTensor, std::decay_t<T>>) {
-      return domain == t.domain;
-    } else {
-      return domain == std::forward<T>(t);
+    if constexpr (std::is_same_v<
+                      AbstractTensorWithInfo<Info>,
+                      std::decay_t<T>>) {
+      return domain == t.domain && info == t.info;
     }
+    return false;
   }
 
   template <typename T>
   bool operator!=(T&& t) const {
     return !operator==(std::forward<T>(t));
   }
 
-  AbstractTensor& parallelize(int64_t axis, ParallelType parallel_type);
+  AbstractTensorWithInfo<Info>& parallelize(
+      int64_t axis,
+      ParallelType parallel_type) {
+    axis = wrapDim(axis, (int64_t)domain.size());
+    AbstractId::dispatch(DispatchParallelize{}, parallel_type, domain[axis]);
+    return *this;
+  }
+
+  AbstractTensorWithInfo<Info>& split(
+      int64_t axis,
+      Val* factor,
+      bool inner_split = true) {
+    axis = wrapDim(axis, (int64_t)domain.size());
+    auto [outer, inner] = AbstractId::dispatch(
+        DispatchSplit{}, domain[axis], factor, inner_split);
+    std::swap(domain[axis], inner);
+    domain.insert(domain.begin() + axis, outer);
+
+    info.insert(info.begin() + axis, info[axis]);
```

Comment:
Do we need a `Info::split`?

Reply:
It is easy to add that, but since generic copy is sufficient for the matmul case I didn't add an interface for all of the scheduling primitives. However, I think it's probably clearer to do so. Especially since that gives me a place to document what we do for swizzle (treated same as merge in `TagSetInfo`).